src/dp.rs

```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Differential privacy (DP) primitives.
+use crate::field::FieldElement;
+use crate::vdaf::prg::Seed;
+use std::fmt::Debug;
+
+// TODO(tholop): We could also use pairs of integers, or even
+// floats to store the DiscreteGaussian DP parameters?
+// Using BigInts might make it harder to communicate `DiscreteGaussian` to the Collector.
+use num_rational::BigRational;
+
+/// DP-related errors.
+#[derive(Debug, thiserror::Error)]
+pub enum DpError {
+    /// Error while sampling from a specific noise distribution
+    #[error("sampling error")]
+    SamplingError,
+}
+
+/// Trait for a noise distribution over a (vector space over a) finite field.
+/// Useful for differential privacy.
+/// Inherited by specific mechanism instantiations such as `[DiscreteGaussian]`
+/// that will enforce the VDAF-dependent DP guarantees.
+pub trait Dist: Clone + Debug {
+    /// Sample `len` elements from the distribution over `F`.
+    /// The randomness comes from a Prg seeded by `seed`.
+    /// TODO(tholop): do we need a generic SEED_SIZE here, or 16 is fine?
+    fn sample<F: FieldElement, const SEED_SIZE: usize>(
+        &self,
+        seed: Seed<SEED_SIZE>,
+        len: usize,
+    ) -> Result<Vec<F>, DpError>;
+
+    // TODO(tholop): do we need another method to send the distribution parameters
+    // to the Collector?
+}
+
+/// Discrete Gaussian noise distribution with mean 0 and standard deviation `sigma`.
+/// `sigma` can be computed as a function of DP parameters `epsilon`, `delta`
+/// and `sensitivity`. `sensitivity` depends on the VDAF and potentially the number of measurements.
+/// `sigma` is sufficient to characterize the distribution, but other parameters
+/// are kept for interpretability.
+///
+/// The distribution is defined over Z, represented by BigInts
+/// To implement `Dist`, the `sample` method can project on a finite F if necessary,
+/// as long as the DP guarantees are the same as the mechanism over Z
+/// (the proof is implementation dependent, e.g. DP post-processing)
+///
+/// TODO(tholop): depending on how we handle `num_measurements`, rename `sensitivity`
+/// to something like `base_sensitivity` or `sensitivity_when_num_measurements_is_1`,
+/// and explain that `add_noise_to_agg_share` is in charge of scaling the noise properly.
+pub struct DiscreteGaussian {
+    epsilon: BigRational,
+    delta: BigRational,
+    sigma: BigRational,
+    sensitivity: BigRational,
```

Comment:
I think we may need to explain what this is the sensitivity of. Alternately, if this is just the sensitivity of the VDAF's aggregate result with respect to any one valid measurement, then the sensitivity could be computed within `add_noise_to_agg_share()`. If the sensitivity is indeed a dependent variable, then we could split the other parameters apart from the implementation of discrete Gaussian sampling.

Reply:
>  Alternately, if this is just the sensitivity of the VDAF's aggregate result with respect to any one valid measurement, then the sensitivity could be computed within add_noise_to_agg_share()

Yes, I think that the sensitivity of the aggregate result could be computed within `add_noise_to_agg_share`, by using both `num_measurements` and the particular VDAF. It seems cleaner than decomposing the sensitivity into two factors (`num_measurements`-dependent scaling factor in `add_noise_to_agg_share`, and `base_sensitivity` in `DiscreteGaussian`).

I'll remove `sensitivity` from `DiscreteGaussian`, which will then represent an unscaled distribution (sensitivity 1).

Reply:
I realized that we can't always reduce to sensitivity-1 functions by scaling the result or the noise. Indeed, unlike the continuous case, when we work with integers `Delta * N_Z(0, sigma^2)` does not have the same distribution as `N_Z(0, Delta^2 * sigma^2)` for an integer Delta (the two distributions don't even have the same support!). 

Hence I think we should still give a `sensitivity` field to the constructor of the `Distribution` object, where the sensitivity definition and the privacy guarantees are given by the Discrete Gaussian paper.

Reply:
Would it be simpler to make the `DiscreteGaussian` constructor to take only sigma, representing the distribution's standard deviation (or sigma squared, for variance) instead of the sensitivity and the sensitivity-1 value of sigma? Would the constructor otherwise just multiply the `Delta` and `sigma` arguments together? It appears to me that the "Sampling from a Discrete Gaussian" algorithm in section 5.3 only needs a single parameter to describe the probability distribution.

Reply:
The distribution is completely described by a single parameter, but the DP guarantees depend on the sensitivity and the budget (e.g. Thm 4 or Thm 7 of https://arxiv.org/pdf/2004.00010.pdf). In the current draft the constructor takes `epsilon`, `delta` and `Delta`: https://github.com/divviup/libprio-rs/blob/d22d271c54f716507b3b5573bfd7da103a29d49e/src/dp.rs#L59 (and not `sigma` and `Delta`), and is supposed to compute the right value of `sigma`. 

Are you suggesting to construct `DiscreteGaussian` only from `sigma`, and leave it up to the user how to compute and interpret `sigma` (outside of libprio)? That would simplify some things (e.g. we wouldn't need to provide constructors for every possible accounting method like zCDP, RDP, Approximate DP; and the DP guarantees could be computed when `num_measurements` is known, maybe at the DAP level if we have a fixed batch size). However that would also leave more room for mistakes (e.g. if the user doesn't compute the sensitivity properly, although we could still provide some optional convenience functions to help computing sigma).

Reply:
We could still handle this in libprio, but I think separating higher-level DP mechanisms from lower-level probability distributions while doing so would be valuable. Here's one way we could arrange it, with separate traits and types.

```rust
pub trait Distribution: Clone + Debug {
    fn sample<F: FieldElement, S: SeedStream>(&self, seed_stream: &mut S, noise: &mut [F]);
}

/// N_Z(0, sigma^2)
pub struct DiscreteGaussian {
    sigma: Rational,
}

impl DiscreteGaussian {
    pub fn new(sigma :Rational) -> Self {
        Self {
            sigma,
        }
    }
}

impl Distribution for DiscreteGaussian {
    fn sample<F: FieldElement, S: SeedStream>(&self, seed_stream: &mut S, noise: &mut [F]) {
        todo!()
    }
}

/// Marker trait for differential privacy mechanisms.
pub trait PrivacyMechanism {}

struct ZeroConcentratedDifferentialPrivacy {
    epsilon: Rational,
}

impl ZeroConcentratedDifferentialPrivacy {
    pub fn new(epsilon: Rational) -> Self {
        Self { epsilon }
    }
}

pub trait AggregatorWithNoise<const VERIFY_KEY_SIZE: usize, const NONCE_SIZE: usize, DP>:
    Aggregator<VERIFY_KEY_SIZE, NONCE_SIZE>
where DP: PrivacyMechanism,
{
    fn add_noise_to_agg_share(
        &self,
        mechanism: &DP,
        agg_param: &Self::AggregationParam,
        agg_share: &mut Self::AggregateShare,
        num_measurements: usize,
    ) -> Result<(), VdafError>;
}

impl AggregatorWithNoise<16, 16, ZeroConcentratedDifferentialPrivacy> for Prio3Count {
    fn add_noise_to_agg_share(
        &self,
        mechanism: &ZeroConcentratedDifferentialPrivacy,
        agg_param: &Self::AggregationParam,
        agg_share: &mut Self::AggregateShare,
        num_measurements: usize,
    ) -> Result<(), VdafError> {
        let sensitivity = Rational::new(1, 1); // differs per VDAF (and sensitivity definition)
        let sigma = sensitivity / mechanism.epsilon;
        let dg = DiscreteGaussian::new(sigma);

        let mut noise = [Field64::zero()];
        let seed = Seed::generate()?; // TODO: make seed an explicit parameter for testability.
        let mut seed_stream = PrgSha3::seed_stream(&seed, &[], &[]);
        dg.sample(&mut seed_stream, &mut noise);

        agg_share.sum(&noise)
    }
}
```

This arrangement gives us static dispatch on the VDAF (and thus the query function) and the DP mechanism at the same time. Do you think this would be a complete enough API to model what we need?

Reply:
This looks good to me.

Reply:
I think this is complete enough and I implemented it. Thanks for the idea! I just used  `DifferentialPrivacyBudget` instead of `PrivacyMechanism`, since this struct only tracks an epsilon (or epsilon-delta, or rho, etc.) and not a full mechanism (which would be a full noise-adding algorithm).